client/telemetry/src/lib.rs

```diff
@@ -333,61 +372,90 @@ impl TelemetryHandle {
 	///
 	/// The `connection_message` argument is a JSON object that is sent every time the connection
 	/// (re-)establishes.
-	pub fn start_telemetry(
-		&mut self,
-		span: TelemetrySpan,
-		endpoints: TelemetryEndpoints,
-		connection_message: ConnectionMessage,
-	) -> TelemetryConnectionNotifier {
-		let Self { message_sender } = self;
+	pub fn start_telemetry(&mut self, connection_message: ConnectionMessage) {
+		let Self {
+			message_sender: _,
+			register_sender,
+			id,
+			connection_notifier: _,
+			endpoints,
+		} = self;
 
-		let connection_notifier = TelemetryConnectionNotifier {
-			message_sender: message_sender.clone(),
-			addresses: endpoints.0.iter().map(|(addr, _)| addr.clone()).collect(),
+		let endpoints = match endpoints.take() {
+			Some(x) => x,
+			None => {
+				error!(
+					target: "telemetry",
+					"This telemetry instance has already been initialized!",
+				);
+				return;
+			}
 		};
 
-		match span.0.id() {
-			Some(id) => {
-				match message_sender.unbounded_send(Register::Telemetry {
-					id,
-					endpoints,
-					connection_message,
-				}) {
-					Ok(()) => {}
-					Err(err) => error!(
-						target: "telemetry",
-						"Could not initialize telemetry: \
-						the telemetry is probably already running: {}",
-						err,
-					),
-				}
-			}
-			None => error!(
+		match register_sender.unbounded_send(Register::Telemetry {
+			id: *id,
+			endpoints,
+			connection_message,
+		}) {
+			Ok(()) => {}
+			Err(err) => error!(
 				target: "telemetry",
-				"Could not initialize telemetry: the span could not be entered",
+				"Could not initialize telemetry: \
+				the telemetry is probably already running: {}",
+				err,
 			),
 		}
+	}
+
+	/// Make a new clonable handle to this [`Telemetry`]. This is used for reporting telemetries.
+	pub fn handle(&self) -> TelemetryHandle {
+		TelemetryHandle {
+			message_sender: self.message_sender.clone(),
+			id: self.id,
+			connection_notifier: self.connection_notifier.clone(),
+		}
+	}
+}
+
+/// Handle to a [`Telemetry`]. Used to report telemetries.
+#[derive(Debug, Clone)]
+pub struct TelemetryHandle {
+	message_sender: mpsc::Sender<TelemetryMessage>,
+	id: Id,
+	connection_notifier: TelemetryConnectionNotifier,
+}
+
+impl TelemetryHandle {
+	/// Send telemetries.
+	pub fn send(&mut self, verbosity: VerbosityLevel, payload: TelemetryPayload) {
+		match self.message_sender.try_send((self.id, verbosity, payload)) {
+			Ok(()) => {}
+			Err(err) if err.is_full() => todo!("overflow"),
+			Err(_) => unreachable!(),
+		}
+	}
 
-		connection_notifier
+	/// Get event stream for telemetry connection established events.
+	///
+	/// This function will return an error if the telemetry has already been started by
+	/// [`TelemetryWorkerHandle::start_telemetry`].
+	pub fn on_connect_stream(&self) -> TracingUnboundedReceiver<()> {
```

Comment:
Why does this stream needs to be unbounded?

If the stream is full, you can just drop `on_connect` messages.

Reply:
It's historical. I didn't change this of the previous implementation. But I agree with you. Should I include this change in this PR or in a follow up? (There is no bounded version of this channel atm)

Reply:
The change isn't that big, so just change it here.

Reply:
Done in 81f7df8207